File: app/src/main/java/com/dcrandroid/fragments/PoliteiaFragment.java
package com.dcrandroid.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.Toast;

import com.dcrandroid.R;
import com.dcrandroid.adapter.ProposalAdapter;
import com.dcrandroid.data.Proposal;
import com.dcrandroid.util.NetworkTask;

import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

public class PoliteiaFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {
    private List<String> filters = new ArrayList<>();
    private ArrayAdapter<String> spinnerAdapter;
    private ArrayList<Proposal> proposals = new ArrayList<>();
    private ArrayList<Proposal> filtered_proposals = new ArrayList<>();
    private RecyclerView recyclerView;
    private ProposalAdapter proposalAdapter;
    private SwipeRefreshLayout swipeRefreshLayout;
    private Spinner spinner;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_politeia, container, false);

        recyclerView = view.findViewById(R.id.recycler_view);
        swipeRefreshLayout = view.findViewById(R.id.swipe_refresh_layout);


        spinner = view.findViewById(R.id.spinnerFilter);


        filters.add("All");

        spinnerAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, filters);
        spinnerAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(spinnerAdapter);


        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                filter_proposals();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        swipeRefreshLayout.setOnRefreshListener(this);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        proposalAdapter = new ProposalAdapter(filtered_proposals, getContext());
        recyclerView.setAdapter(proposalAdapter);

        swipeRefreshLayout.post(
                new Runnable() {
                    @Override
                    public void run() {
                        loadProposals();
                    }
                }
        );

        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() == null || getContext() == null) {
            return;
        }
        getActivity().setTitle(getString(R.string.politeia));
    }

    void loadProposals() {
        swipeRefreshLayout.setRefreshing(true);
        final HashMap<String, String> config = new HashMap<>();
        config.put("url", getString(R.string.politeia_proposals_url));
        new NetworkTask(config, new NetworkTask.AsyncResponse() {
            @Override
            public void onResponse(@NotNull final String proposalsResponse) {
                HashMap<String, String> config = new HashMap<>();
                config.put("url", "https://proposals.decred.org/api/v1/proposals/votestatus");
                new NetworkTask(config, new NetworkTask.AsyncResponse() {
                    @Override
                    public void onResponse(@NotNull String votesReponse) {
                        try {

                            swipeRefreshLayout.setRefreshing(false);
                            JSONObject jsonObject1 = new JSONObject(proposalsResponse);
                            JSONArray jsonProposals = jsonObject1.getJSONArray("proposals");
                            if (jsonProposals != null) {
                                proposals.clear();
                                int abandoned = 0;
                                for (int i = 0; i < jsonProposals.length(); i++) {
                                    JSONObject jsonProposal = jsonProposals.getJSONObject(i);
                                    Proposal proposal = new Proposal();
                                    proposal.setName(jsonProposal.getString("name"));
                                    proposal.setState(jsonProposal.getInt("state"));
                                    proposal.setStatus(jsonProposal.getInt("status"));
                                    proposal.setTimestamp(jsonProposal.getLong("timestamp"));
                                    proposal.setUserid(jsonProposal.getString("userid"));
                                    proposal.setUsername(jsonProposal.getString("username"));
                                    proposal.setPublickey(jsonProposal.getString("publickey"));
                                    proposal.setSignature(jsonProposal.getString("signature"));
                                    proposal.setVersion(jsonProposal.getString("version"));
                                    proposal.setNumcomments(jsonProposal.getInt("numcomments"));

                                    if (jsonProposal.getInt("status") == 6) abandoned++;

                                    JSONArray jsonFiles = jsonProposal.getJSONArray("files");
                                    ArrayList<Proposal.File> files = new ArrayList<>();

                                    for (int j = 0; j < jsonFiles.length(); j++) {
                                        JSONObject jsonFile = jsonFiles.getJSONObject(j);
                                        Proposal.File file = new Proposal().new File();
                                        file.setName(jsonFile.getString("name"));
                                        file.setMime(jsonFile.getString("mime"));
                                        file.setDigest(jsonFile.getString("digest"));
                                        file.setPayload(jsonFile.getString("payload"));
                                        files.add(file);
                                    }

                                    Proposal.CensorshipRecord record = new Proposal().new CensorshipRecord();
                                    record.setToken(jsonProposal.getJSONObject("censorshiprecord").getString("token"));
                                    record.setMerkle(jsonProposal.getJSONObject("censorshiprecord").getString("merkle"));
                                    record.setSignature(jsonProposal.getJSONObject("censorshiprecord").getString("signature"));
                                    proposal.setFiles(files);
                                    proposal.setCensorshipRecord(record);
                                    proposals.add(proposal);
                                }


                                JSONObject jsonObject = new JSONObject(votesReponse);
                                JSONArray votes = jsonObject.getJSONArray("votesstatus");
                                int pre = 0, active = 0, finished = 0;
                                for (int i = 0; i < proposals.size(); i++) {
                                    for (int j = 0; votes != null && j < votes.length(); j++) {
                                        JSONObject vote = votes.getJSONObject(j);
                                        if (vote.getString("token").equals(proposals.get(i).getCensorshipRecord().getToken())) {
                                            if (vote.getInt("status") == 1) pre++;
                                            if (vote.getInt("status") == 2) active++;
                                            if (vote.getInt("status") == 3) finished++;
                                            Proposal.VoteStatus voteStatus = new Proposal().new VoteStatus();
                                            voteStatus.setStatus(vote.getInt("status"));
                                            voteStatus.setTotalvotes(vote.getInt("totalvotes"));
                                            if (!vote.isNull("optionsresult")) {
                                                voteStatus.setNo(vote.getJSONArray("optionsresult").getJSONObject(0).getInt("votesreceived"));
                                                voteStatus.setYes(vote.getJSONArray("optionsresult").getJSONObject(1).getInt("votesreceived"));
                                            }
                                            voteStatus.setTotalvotes(vote.getInt("totalvotes"));
                                            proposals.get(i).setVoteStatus(voteStatus);
                                        }

                                    }
                                }

                                if (getActivity() != null) {
                                    final int finalPre = pre, finalActive = active, finalFinished = finished, finalAbandoned = abandoned;
                                    getActivity().runOnUiThread(new Runnable() {
                                        @Override
                                        public void run() {
                                            filters.clear();
                                            filters.add("All (" + proposals.size() + ")");
                                            filters.add("Pre-Voting (" + finalPre + ")");
                                            filters.add("Active Voting (" + finalActive + ")");
                                            filters.add("Finished Voting (" + finalFinished + ")");
                                            filters.add("Abandoned (" + finalAbandoned + ")");
                                            spinnerAdapter.notifyDataSetChanged();
                                            filter_proposals();
                                        }
                                    });
                                }
                            }


                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }

                    @Override
                    public void onFailure(final Throwable t) {
                        t.printStackTrace();
                    }
                }).execute();

            }

            @Override
            public void onFailure(final Throwable t) {
                swipeRefreshLayout.setRefreshing(false);
                t.printStackTrace();
                if (getActivity() != null) {
                    getActivity().runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(getActivity(), t.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
                }
            }
        }).execute();
    }

    void filter_proposals() {
        int pos = spinner.getSelectedItemPosition();
        filtered_proposals.clear();
        switch (pos) {
            case 0:
                filtered_proposals.addAll(proposals);
                break;
            case 1:
            case 2:
            case 3:
                for (int i = 0; i < proposals.size(); i++) {
                    if (proposals.get(i).getVoteStatus() == null) continue;
                    if (proposals.get(i).getVoteStatus().getStatus() == pos) {
                        filtered_proposals.add(proposals.get(i));
                    }
                }
                break;
            case 4:
                for (int i = 0; i < proposals.size(); i++) {
                    if (proposals.get(i).getStatus() == 6) {
                        filtered_proposals.add(proposals.get(i));
                    }
                }

                break;
            default:
                filtered_proposals.addAll(proposals);
        }
        proposalAdapter.notifyDataSetChanged();
    }

    @Override
    public void onRefresh() {
        loadProposals();
    }
}

File: app/src/main/java/com/dcrandroid/fragments/PoliteiaFragment.kt
/*
 * Copyright (c) 2018-2019 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.dcrandroid.BuildConfig
import com.dcrandroid.R
import com.dcrandroid.adapter.ProposalAdapter
import com.dcrandroid.data.Constants
import com.dcrandroid.data.Proposal
import com.dcrandroid.util.Deserializer
import com.dcrandroid.util.PreferenceUtil
import com.dcrandroid.util.QueryAPI
import com.google.gson.GsonBuilder
import org.json.JSONObject
import java.util.ArrayList

const val PRE_VOTING = 1
const val ACTIVE_VOTING = 2
const val FINISHED_VOTING = 3
const val ABANDONED_PROPOSALS = 6

class PoliteiaFragment: Fragment(), SwipeRefreshLayout.OnRefreshListener, QueryAPI.QueryAPICallback {

    private val filters = ArrayList<String>()
    private var spinnerAdapter: ArrayAdapter<String>? = null
    private val proposals = ArrayList<Proposal>()
    private val filteredProposals = ArrayList<Proposal>()
    private var recyclerView: RecyclerView? = null
    private var proposalAdapter: ProposalAdapter? = null
    private var swipeRefreshLayout: SwipeRefreshLayout? = null
    private var spinner: Spinner? = null
    private var util: PreferenceUtil? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_politeia, container, false)

        util = PreferenceUtil(requireContext())

        recyclerView = view.findViewById(R.id.recycler_view)
        swipeRefreshLayout = view.findViewById(R.id.swipe_refresh_layout)

        spinner = view.findViewById(R.id.spinnerFilter)

        filters.add(getString(R.string.all))

        spinnerAdapter = ArrayAdapter(context!!, android.R.layout.simple_spinner_item, filters)
        spinnerAdapter!!.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner!!.adapter = spinnerAdapter

        spinner!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                filterProposals()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {}
        }

        swipeRefreshLayout!!.setOnRefreshListener(this)

        recyclerView!!.layoutManager = LinearLayoutManager(context)
        proposalAdapter = ProposalAdapter(filteredProposals, context!!)
        recyclerView!!.adapter = proposalAdapter

        swipeRefreshLayout!!.post{
            loadProposals()
        }

        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        requireActivity().setTitle(R.string.politeia)
    }

    private fun loadProposals(){
        swipeRefreshLayout!!.isRefreshing = true
        val vettedProposalsUrl = getString(R.string.vetted_proposals_url, BuildConfig.PoliteiaHost)
        val userAgent = util!!.get(Constants.USER_AGENT, Constants.EMPTY_STRING)
        QueryAPI(vettedProposalsUrl, userAgent, this).execute()
    }

    private fun filterProposals(){
        val pos = spinner!!.selectedItemPosition
        filteredProposals.clear()

        when(pos){
            in 1..3 -> filteredProposals.addAll(proposals.filter { it.voteStatus != null && it.voteStatus!!.status == pos })
            4 -> filteredProposals.addAll(proposals.filter { it.status == ABANDONED_PROPOSALS })
            else -> filteredProposals.addAll(proposals)
        }

        proposalAdapter!!.notifyDataSetChanged()
    }

    override fun onRefresh() {
        loadProposals()
    }

    private fun parseResults(proposalsJson: String, voteStatusJson: String){
        val gson = GsonBuilder()
                .registerTypeHierarchyAdapter(ArrayList::class.java, Deserializer.ProposalDeserializer())
                .registerTypeHierarchyAdapter(Proposal.VoteStatus.OptionsResult::class.java, Deserializer.OptionsResultDeserializer())
                .create()

        var parentObj = JSONObject(proposalsJson)
        val tempProposals = gson.fromJson(parentObj.getJSONArray("proposals").toString(), Array<Proposal>::class.java)
        proposals.clear()
        proposals.addAll(tempProposals)

        parentObj = JSONObject(voteStatusJson)
        val voteStatus = gson.fromJson(parentObj.getJSONArray("votesstatus").toString(), Array<Proposal.VoteStatus>::class.java)

        proposals.map { proposal ->
            val status = voteStatus.find { it.token == proposal.censorshipRecord!!.token }
            if(status != null){
                proposal.voteStatus = status
            }
        }

        val preVoting = voteStatus.count { it.status == PRE_VOTING }
        val activeVoting = voteStatus.count { it.status == ACTIVE_VOTING }
        val finishedVoting = voteStatus.count { it.status == FINISHED_VOTING }
        val abandoned = proposals.count { it.status == ABANDONED_PROPOSALS }

        filters.clear()
        filters.add("All (${proposals.size})")
        filters.add("Pre-Voting ($preVoting)")
        filters.add("Active Voting ($activeVoting)")
        filters.add("Finished Voting ($finishedVoting)")
        filters.add("Abandoned ($abandoned)")
        spinnerAdapter!!.notifyDataSetChanged()
        filterProposals()

        swipeRefreshLayout!!.isRefreshing = false
    }

    override fun onQueryAPISuccess(result: String?) {

        val voteStatusUrl = getString(R.string.proposals_vote_status_url, BuildConfig.PoliteiaHost)
        val userAgent = util!!.get(Constants.USER_AGENT, Constants.EMPTY_STRING)
        val proposalsJson = result

        QueryAPI(voteStatusUrl, userAgent, object : QueryAPI.QueryAPICallback{
            override fun onQueryAPISuccess(result: String?) {
                parseResults(proposalsJson!!, result!!)
            }

            override fun onQueryAPIError(e: Exception) {
                e.printStackTrace()
                activity!!.runOnUiThread {
                    swipeRefreshLayout!!.isRefreshing = false
                    Toast.makeText(context, e.localizedMessage, Toast.LENGTH_SHORT).show()
                }
            }
        }).execute()
    }

    override fun onQueryAPIError(e: Exception) {
        e.printStackTrace()
        activity!!.runOnUiThread {
            swipeRefreshLayout!!.isRefreshing = false
            Toast.makeText(context, e.localizedMessage, Toast.LENGTH_SHORT).show()
        }
    }
}